etc/spack/defaults/packages.yaml

```diff
@@ -36,6 +36,10 @@ packages:
       openfoam: [openfoam-com, openfoam-org, foam-extend]
       pil: [py-pillow]
       pkgconfig: [pkgconf, pkg-config]
+      postgresql-c: [postgresql-client, postgresql]
+      postgresql-python: [postgresql-client+python, postgresql+python]
+      postgresql-perl: [postgresql-client+perl, postgresql+perl]
```

Comment:
@chissg There are two interesting precedents here: 

* I wasn't aware Spack supported setting variants on providers (in this file). I don't know all the repercussions of this and I'm pretty sure at least that Spack doesn't unit test this at the moment (perhaps it ought to work and you've likely checked it yourself, but my point is that this is potentially fragile). You mention `tbb` as a precedent and I'll note that the `tbb` provider mentioned here punts on the issue by not including `intel-parallel-studio` (which conditionally provides `tbb`).
* It isn't typical to get this fine-grained with virtuals (e.g. having one for each language binding). I can see though why this appeals since each language-binding requires installation of the language itself. I don't consider the savings in installation time/resources to be a major benefit, but I do think that this saves potential user headaches with configuring/managing each language.

The typical approach in Spack for providing limited functionality (e.g. just the client) would be the `~server` option you mention at https://github.com/spack/spack/pull/7728#issuecomment-396275184. I'm not sure I want to require that approach, but I can say that the above issues require more thinking whereas the `~server` variant would be more-straightforward to review. If you have some time to wait, I can try to check into supporting this PR as-is at some point in the next couple weeks.

I'll also say that one major difference between this and https://github.com/spack/spack/pull/7729 is that the mariadb client is provided as an entirely separate source archive, while the postgres client uses the same source with a different build. In that case I recognize it is more straightforward to maintain the packages separately.

Reply:
@scheibelp Thanks for considering this. I can certainly look into unit tests for the virtual-provider-with-variants requirement above, although I'm really not familiar with that part of the code yet. Since we require substantially more variant-setting on our installations of `postgresql-client` (`libedit` over `readline`, for example), I've been supplying the dependency on the command line with `^` notation. I can verify however, that even though the default value of the `python` variant is `off`, one can say `spack install postgresql-python` and the outcome is the installation of `postgresql-celint+python`. Even `spack info postgresql-client` gives:  
```
...
Virtual Packages: 
    postgresql-client+tcl provides postgresql-tcl
    postgresql-client+python provides postgresql-python
    postgresql-client+perl provides postgresql-perl
    postgresql-client provides postgresql-c
```

 I can say though, that there is apparently a weakness in the dependency resolution if a given dependent package `depends_on()` two virtual packages that both resolve to the same concrete package with different (although in this case not incompatible) variant requirements.

As far as time constraints go, We are operating off a fork containing quite a few things that haven't been fed back yet, so we're OK for a while. Obviously the fewer material differences the better of course, and once this is resolved one way or the other I can contribute some of our improvements to the Root recipe (related: PR #8428 and comments).